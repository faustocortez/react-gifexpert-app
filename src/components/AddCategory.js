import React, { useState } from 'react'
import PropTypes  from 'prop-types';

export const AddCategory = ({ setCategories }) => {
    
    const [inputValue, setInputValue] = useState(''); // initialze value to aviod undefinied
    
    const handleInputChange = (e) => {
        const text = e.target.value;
        setInputValue(text);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        

        if (inputValue.trim().length > 2) {
            // We can use the previos data in categories only passing the setCategory function
            setCategories(categories => [inputValue, ...categories]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <p> { inputValue } </p>
            <input
                type="text"
                onChange={ handleInputChange }
                value={inputValue}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}
