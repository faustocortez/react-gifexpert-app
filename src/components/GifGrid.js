import React, { useState, useEffect } from 'react'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const [images, setImages] = useState([]);

    // component did mount in react classes
    // the second parameter "[]" is a list of dependencies
    // if [] equal to one execution
    useEffect(() => {
        getGifs();
    }, []);

    const getGifs = async() => {
        const url = 'https://api.giphy.com/v1/gifs/search?api_key=LorHcp1hTRLEIFFnzCsCcrXK0nOAS0Ux&q=Rick+and+Morty&limit=10';
        const response = await fetch( url );
        const { data } = await response.json();

        const gifs = data.map(img => {
            const {
                id,
                title,
                images: { downsized_medium: { url} },
            } = img;
            return  {
                id,
                title,
                url
            }
        })

        console.log(gifs);
        setImages(gifs);
    }

    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                { 
                    images.map( img => (    
                        <GifGridItem
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            </div>
        </>
    )
}
