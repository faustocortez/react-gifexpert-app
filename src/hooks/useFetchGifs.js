import { useState, useEffect } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = ( category ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true,
    });

    useEffect(() => {

        getGifs(category)
            .then(imgs => {
                setTimeout(() => {
                    setState({
                        data: imgs,
                        loading: false,
                    });
                }, 1500);
            })

    }, [category])

    return state;
}

// component did mount in react classes
// the second parameter "[]" is a list of dependencies
// if [] equal to one execution
// useEffect(() => {
//     getGifs(category)
//         // como el argumente del then es el mismo mandado en la función
//         // interna se usa shortand la función interna de manera directa.
//         // antes: then(imgs => setImages(imgs))
//         .then(setImages);
// }, [ category ]);

// ** LOS EFECTOS NO PUEDEN SER ASYNC PORQUE ESPERAN ALGO SÍNCRONO.
