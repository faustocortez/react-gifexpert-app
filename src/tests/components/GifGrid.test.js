import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
// para fingir la llamada al archivo
jest.mock('../../hooks/useFetchGifs');


describe('Pruebas del componente GifGrid', () => {
    const category = 'Demon slayer';
    const mockUseFetchGifs = useFetchGifs;
    
    test('* renderizar el componente por primera vez correctamente', () => {
        
        // valores del estado inicial del useFetch.
        mockUseFetchGifs.mockImplementation( category => {
            return { data: [], loading: true }
        });

        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot(); 
    });

    test('* debe de mostrar items cuando se cargan imÃ¡genes useFetch', () => {
        
        // aplicando mocks: como si el componente ya cuenta con la inforamacion del request

        const data = [{
            id: '0',
            url: 'https://lorem.com/img',
            title: 'mockTitle'
        }];
    
        mockUseFetchGifs.mockImplementation( category => {
            return { data, loading: false };
        });
    
        const wrapper = shallow( <GifGrid category={ category } /> );
        expect( wrapper ).toMatchSnapshot();
        
        
    });
    
    
});
