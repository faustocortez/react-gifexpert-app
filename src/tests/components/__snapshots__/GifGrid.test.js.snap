// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente GifGrid * debe de mostrar items cuando se cargan im√°genes useFetch 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Demon slayer
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="0"
      key="0"
      title="mockTitle"
      url="https://lorem.com/img"
    />
  </div>
</Fragment>
`;

exports[`Pruebas del componente GifGrid * renderizar el componente por primera vez correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Demon slayer
  </h3>
  <p
    className=" animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
